### ISSUE TITLE
[BUG] kakoune used in aerc cause redrawing issue

### ISSUE THREAD

**Issue by stacyharper** [mawww/kakoune#4689]
### Version of Kakoune

2021.11.08-r1 (Alpine Linux)

### Reproducer

Setup EDITOR=kak and use a basic aerc config. Try to reply-quote to a mail

### Outcome

Sometime redrawing is completly broken: Here a short vod that show the issue on the  third tentative.

https://tube.missbanal.net/w/e4ab1d4d-73b9-442c-99ea-9aeb234631e7

### Expectations

Kakoune shoult redraw itself correctly while embeded

### Additional information

I'm on Alpine Linux Edge and up to date (22-07-28).

I already asked aerc guys and they guess it is kakoune related cause the issue is yet unknown with other editors (vim, nano, etc).

**Comment by 0x5424** [comment]
Though I'm not using the aerc client, I'm experiencing similar after building from source (commit hash `360a6847be5c`).

**Distro**: Ubuntu server 20.04.5

It's not present in the package published in `apt`, which is quite old (A build from December 2019)

Attached a short clip before & after starting a tmux session. I can also confirm xterm works fine
https://user-images.githubusercontent.com/25315226/195968296-fc93363b-bfd7-45d4-9a7f-811ef6ad8da4.MOV

**Comment by Screwtapello** [comment]
You say xterm works fine, and tmux (which is technically a terminal, as far as Kakoune is concerned) works fine. Which terminal causes problems, and does it still cause problems if you launch Kakoune without its configuration (`kak -n`)?

**Comment by 0x5424** [comment]
It's a bit puzzling as I've tested a few different configurations... but I've identified the problem to be the [SSH client I used to test on mobile](https://termius.com/). My apologies, as it may be more appropriate to report to their issue tracker instead.

---

> does it still cause problems if you launch Kakoune without it's configuration

The issue still appeared, with or without the kakrc. Unfortunately, the ssh client is a freemium closed-source product so I cannot provide any additional debug info.

In the event it somehow helps the Kakoune team with the titular issue, I've attached a few screenshots.
1. The first is running `kak -n`.
2. The second uses a kakrc with `set-option global ui_options terminal_status_on_top=true`.
3. The third screenshot is the terminal window after quitting a kakoune client which used `terminal_status_on_top=true`. It prints the final state of the status bar before quitting (notice the `:q` )
4. A one-off error I wasn't able to reproduce again, wherein mojibake prints to line 2 using the same `Information` face from the changelog. This also had `terminal_status_on_top=true`

<details><summary>Terminal status on bottom (kak -n)</summary>
<img src="https://user-images.githubusercontent.com/25315226/196090960-1e723f1f-2dfe-477b-b299-8de7ac934898.PNG"/>
</details>

<details><summary>Terminal status on top</summary>
<img src="https://user-images.githubusercontent.com/25315226/196091234-e06b0138-0580-4491-abab-9fb45f92833c.PNG"/>
</details>

<details><summary>After closing a session with terminal status on top</summary>
<img src="https://user-images.githubusercontent.com/25315226/196091245-a6a9a5b7-97af-4005-9f3c-0bf469d4bbda.PNG"/>
</details>

<details><summary>Mojibake following changelog</summary>
<img src="https://user-images.githubusercontent.com/25315226/196092900-52079d1f-c86a-458d-be20-f516d41cd3b6.PNG"/>
</details>

The client only seems to only draw the final line of the ui. Hope this may be of some use.

**Comment by Screwtapello** [comment]
Unfortunately, terminal technology is in a similar place to the where HTML and browsers were before HTML5: there's a great number of independently-implemented producers and consumers, but no solid community-accepted interoperable standard. Since there's no standard to test against, each producer only tests against the five (or so) most common consumers, and each consumer only tests against the five (or so) most common producers. As a result, there's a lot of scope for somebody to say "I'll just cut *this* corner and I'm sure it'll be fine" and it *is* fine until somebody tries to run a niche editor in a niche terminal emulator and all hell breaks loose.

Clearly Kakoune is doing something that confuses Termius, but I've never seen that particular behaviour before and it's hard to guess what the problem might be. If I had to investigate further, my plan would be:

1. use the `script` tool to run Kakoune and capture all of its output, including escape sequences, to a file:

        script -c "kak -n" output-log.txt

2. When Kakoune appears, verify the output is still glitchy
3. quit Kakoune with `:q`
4. `cat output-log.txt` to verify that it more-or-less produces the same result
5. start bisecting `output-log.txt` to find some escape sequence where up to that point, `cat`ting the file produces reasonable output, but after that point it glitches

It'll help to recognise the escape sequences Kakoune produces on startup to configure the terminal:

https://github.com/mawww/kakoune/blob/287217b98715faaddab019f4e17ab54fccaaec70/src/terminal_ui.cc#L1414-L1423

...and also [a list of control sequences most terminals understand](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Controls-beginning-with-ESC).

Also, since you mention mojibake, make sure that you're using a UTF-8 locale (run the `locale` command and verify each category ends in `.UTF-8` or `.utf8`) before starting Kakoune.

**Comment by stacyharper** [comment]
> You say xterm works fine, and tmux (which is technically a terminal, as far as Kakoune is concerned) works fine. Which terminal causes problems, and does it still cause problems if you launch Kakoune without its configuration (`kak -n`)?

No I said xterm or foot, tmux or not, all of them have the issue.

**Comment by stacyharper** [comment]
Here the content of the mail :

```
On Mon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:                                                                                                        
> Unfortunately, terminal technology is in a similar place to the where HTML and browsers were before HTML5: there's a great number of independently-implemented
>                                                                                                                                                               
> Clearly Kakoune is doing something that confuses Termius, but I've never seen that particular behaviour before and it's hard to guess what the problem might b
>                                                                                                                                                               
> 1. use the `script` tool to run Kakoune and capture all of its output, including escape sequences, to a file:                                                 
>                                                                                                                                                               
>         script -c "kak -n" output-log.txt                                                                                                                     
>                                                                                                                                                               
> 2. When Kakoune appears, verify the output is still glitchy                                                                                                   
> 3. quit Kakoune with `:q`                                                                                                                                     
> 4. `cat output-log.txt` to verify that it more-or-less produces the same result                                                                               
> 5. start bisecting `output-log.txt` to find some escape sequence where up to that point, `cat`ting the file produces reasonable output, but after that point i
>                                                                                                                                                               
> It'll help to recognise the escape sequences Kakoune produces on startup to configure the terminal:                                                           
>                                                                                                                                                               
> https://github.com/mawww/kakoune/blob/287217b98715faaddab019f4e17ab54fccaaec70/src/terminal_ui.cc#L1414-L1423                                                 
>                                                                                                                                                               
> ...and also [a list of control sequences most terminals understand](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Controls-beginning-with-ESC).  
>                                                                                                                                                               
> Also, since you mention mojibake, make sure that you're using a UTF-8 locale (run the `locale` command and verify each category ends in `.UTF-8` or `.utf8`) b
>                                                                                                                                                               
> --                                                                                                                                                            
> Reply to this email directly or view it on GitHub:                                                                                                            
> https://github.com/mawww/kakoune/issues/4689#issuecomment-1280381424                                                                                          
> You are receiving this because you authored the thread.                                                                                                       
>                                                                                                                                                               
```

I typed `l` until the highlighted char is `1` as in `On Mon Oct 17, 2022` and it started to glitch. Here the typescript file as it it just after the last `l` keybind trigger.

```
Script started on 2022-10-17 11:02:43+02:00 [COMMAND="kak -n /tmp/aerc-compose-3282581247.eml" TERM="foot" TTY="/dev/pts/6" COLUMNS="160" LINES="37"]
[?1049h[?1004h[>4;1m[>5u[22t[?25l=[?2026$p[?1006h[?1000h[?1002h[1H[K[2H[K[3H[K[4H[K[5H[K[6H[K[7H[K[8H[K[9H[K[10H[K[11H[K[12H[K[13H[K[14H[K[15H[K[16H[K[17H[K[18H[K[19H[K[20H[K[21H[K[22H[K[23H[K[24H[K[25H[K[26H[K[27H[K[28H[K[29H[K[30H[K[31H[K[32H[K[33H[K[34H[K[35H[K[36H[K[37H[K[1;1H]2;/tmp/aerc-compose-3282581247.eml 1:1  1 sel - client0@[15027] - Kakoune[m[1H[30;47mO[39;49mn Mon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[2H> Unfortunately, terminal technology is in a similar place to the where HTML and browsers were before HTML5: there's a great number of independently-implemented[3H>[K[4H> Clearly Kakoune is doing something that confuses Termius, but I've never seen that particular behaviour before and it's hard to guess what the problem might b[5H>[K[6H> 1. use the `script` tool to run Kakoune and capture all of its output, including escape sequences, to a file:[K[7H>[K[8H>         script -c "kak -n" output-log.txt[K[9H>[K[10H> 2. When Kakoune appears, verify the output is still glitchy[K[11H> 3. quit Kakoune with `:q`[K[12H> 4. `cat output-log.txt` to verify that it more-or-less produces the same result[K[13H> 5. start bisecting `output-log.txt` to find some escape sequence where up to that point, `cat`ting the file produces reasonable output, but after that point i[14H>[K[15H> It'll help to recognise the escape sequences Kakoune produces on startup to configure the terminal:[K[16H>[K[17H> https://github.com/mawww/kakoune/blob/287217b98715faaddab019f4e17ab54fccaaec70/src/terminal_ui.cc#L1414-L1423[K[18H>[K[19H> ...and also [a list of control sequences most terminals understand](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Controls-beginning-with-ESC).  [20H>[K[21H> Also, since you mention mojibake, make sure that you're using a UTF-8 locale (run the `locale` command and verify each category ends in `.UTF-8` or `.utf8`) b[22H>[K[23H> -- [K[24H> Reply to this email directly or view it on GitHub:[K[25H> https://github.com/mawww/kakoune/issues/4689#issuecomment-1280381424[K[26H> You are receiving this because you authored the thread.[K[27H>[K[28H> Message ID: <mawww/kakoune/issues/4689/1280381424@github.com>[K[29H[K[30H[34m~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:1  [34m1 sel[36m - client0@[15027][1;1H]2;/tmp/aerc-compose-3282581247.eml 1:1  1 sel - client0@[15027] - Kakoune[m[1H[30;47mO[39;49mn Mon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[2H> Unfortunately, terminal technology is in a similar place to the where HTML and browsers were before HTML5: there's a great number of independently-implemented[3H>[K[4H> Clearly Kakoune is doing something that confuses Termius, but I've never seen that particular behaviour before and it's hard to guess what the problem might b[5H>[K[6H> 1. use the `script` tool to run Kakoune and capture all of its output, including escape sequences, to a file:[K[7H>[K[8H>         script -c "kak -n" output-log.txt[K[9H>[K[10H> 2. When Kakoune appears, verify the output is still glitchy[K[11H> 3. quit Kakoune with `:q`[K[12H> 4. `cat output-log.txt` to verify that it more-or-less produces the same result[K[13H> 5. start bisecting `output-log.txt` to find some escape sequence where up to that point, `cat`ting the file produces reasonable output, but after that point i[14H>[K[15H> It'll help to recognise the escape sequences Kakoune produces on startup to configure the terminal:[K[16H>[K[17H> https://github.com/mawww/kakoune/blob/287217b98715faaddab019f4e17ab54fccaaec70/src/terminal_ui.cc#L1414-L1423[K[18H>[K[19H> ...and also [a list of control sequences most terminals understand](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Controls-beginning-with-ESC).  [20H>[K[21H> Also, since you mention mojibake, make sure that you're using a UTF-8 locale (run the `locale` command and verify each category ends in `.UTF-8` or `.utf8`) b[22H>[K[23H> -- [K[24H> Reply to this email directly or view it on GitHub:[K[25H> https://github.com/mawww/kakoune/issues/4689#issuecomment-1280381424[K[26H> You are receiving this because you authored the thread.[K[27H>[K[28H> Message ID: <mawww/kakoune/issues/4689/1280381424@github.com>[K[29H[K[30H[34m~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:1  [34m1 sel[36m - client0@[15027][1;1H]2;/tmp/aerc-compose-3282581247.eml 1:2  1 sel - client0@[15027] - Kakoune[1H[39mO[30;47mn[39;49m Mon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:2  [34m1 sel[36m - client0@[15027][1;2H]2;/tmp/aerc-compose-3282581247.eml 1:3  1 sel - client0@[15027] - Kakoune[1H[39mOn[30;47m [39;49mMon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:3  [34m1 sel[36m - client0@[15027][1;3H]2;/tmp/aerc-compose-3282581247.eml 1:4  1 sel - client0@[15027] - Kakoune[1H[39mOn [30;47mM[39;49mon Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:4  [34m1 sel[36m - client0@[15027][1;4H]2;/tmp/aerc-compose-3282581247.eml 1:5  1 sel - client0@[15027] - Kakoune[1H[39mOn M[30;47mo[39;49mn Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:5  [34m1 sel[36m - client0@[15027][1;5H]2;/tmp/aerc-compose-3282581247.eml 1:6  1 sel - client0@[15027] - Kakoune[1H[39mOn Mo[30;47mn[39;49m Oct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:6  [34m1 sel[36m - client0@[15027][1;6H]2;/tmp/aerc-compose-3282581247.eml 1:7  1 sel - client0@[15027] - Kakoune[1H[39mOn Mon[30;47m [39;49mOct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:7  [34m1 sel[36m - client0@[15027][1;7H]2;/tmp/aerc-compose-3282581247.eml 1:8  1 sel - client0@[15027] - Kakoune[1H[39mOn Mon [30;47mO[39;49mct 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:8  [34m1 sel[36m - client0@[15027][1;8H]2;/tmp/aerc-compose-3282581247.eml 1:9  1 sel - client0@[15027] - Kakoune[1H[39mOn Mon O[30;47mc[39;49mt 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[99C/tmp/aerc-compose-3282581247.eml 1:9  [34m1 sel[36m - client0@[15027][1;9H]2;/tmp/aerc-compose-3282581247.eml 1:10  1 sel - client0@[15027] - Kakoune[1H[39mOn Mon Oc[30;47mt[39;49m 17, 2022 at 9:01 AM CEST, Screwtapello wrote:[K[37H[36m[K[98C/tmp/aerc-compose-3282581247.eml 1:10  [34m1 sel[36m - client0@[15027][1;10H
```

**Comment by stacyharper** [comment]
I'm trying to dissect all of this (with another mail content). It looks like this exact terminal code is exactly what we expect from kakoune:

```
[?1049h[?1004h[>4;1m[>5u[22t[?25l=[?2026$p[?1006h[?1000h[?1002h[1H[K[2H[K[3H[K[4H[K[5H[K[6H[K[7H[K[8H[K[9H[K[10H[K[11H[K[12H[K[13H[K[14H[K[15H[K[16H[K[17H[K[18H[K[19H[K[20H[K[21H[K[22H[K[23H[K[24H[K[25H[K[26H[K[27H[K[28H[K[29H[K[30H[K[31H[K[32H[K[33H[K[34H[K[35H[K[36H[K[37H[K[1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:2  1 sel - client0@[13450] - Kakoune[1H[39mO[30;47mn[39;49m Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:2  [34m1 sel[36m - client0@[13450][1;2H]2;/tmp/aerc-compose-4230040883.eml 1:3  1 sel - client0@[13450] - Kakoune[1H[39mOn[30;47m [39;49mSat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:3  [34m1 sel[36m - client0@[13450][1;3H]2;/tmp/aerc-compose-4230040883.eml 1:4  1 sel - client0@[13450] - Kakoune[1H[39mOn [30;47mS[39;49mat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:4  [34m1 sel[36m - client0@[13450][1;4H]2;/tmp/aerc-compose-4230040883.eml 1:5  1 sel - client0@[13450] - Kakoune[1H[39mOn S[30;47ma[39;49mt Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:5  [34m1 sel[36m - client0@[13450][1;5H]2;/tmp/aerc-compose-4230040883.eml 1:6  1 sel - client0@[13450] - Kakoune[1H[39mOn Sa[30;47mt[39;49m Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:6  [34m1 sel[36m - client0@[13450][1;6H]2;/tmp/aerc-compose-4230040883.eml 1:7  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat[30;47m [39;49mOct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:7  [34m1 sel[36m - client0@[13450][1;7H]2;/tmp/aerc-compose-4230040883.eml 1:8  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat [30;47mO[39;49mct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:8  [34m1 sel[36m - client0@[13450][1;8H]2;/tmp/aerc-compose-4230040883.eml 1:9  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat O[30;47mc[39;49mt 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:9  [34m1 sel[36m - client0@[13450][1;9H]2;/tmp/aerc-compose-4230040883.eml 1:10  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oc[30;47mt[39;49m 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:10  [34m1 sel[36m - client0@[13450][1;10H]2;/tmp/aerc-compose-4230040883.eml 1:11  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oct[30;47m [39;49m15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:11  [34m1 sel[36m - client0@[13450][1;11H]2;/tmp/aerc-compose-4230040883.eml 1:12  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oct [30;47m1[39;49m5, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:12  [34m1 sel[36m - client0@[13450]
```

I stripped everything until my `:` char to `:wq` (that show the auto-completion popup and clippy etc).

Cating this give exactly what we want so I dont really know what to do next o_O

![screenshot of the mail rendered in a term](https://dav.missbanal.net/af645652-9e4f-4337-862b-dfe9586ea30d.png)

**Comment by stacyharper** [comment]
I think I bisected the script. Here two typescripts. The first one works great and show the cursor as:

![screenshot working](https://dav.missbanal.net/7685884f-61e0-4898-b5a8-a89fff538bb7.png)

```
[?1049h[?1004h[>4;1m[>5u[22t[?25l=[?2026$p[?1006h[?1000h[?1002h[1H[K[2H[K[3H[K[4H[K[5H[K[6H[K[7H[K[8H[K[9H[K[10H[K[11H[K[12H[K[13H[K[14H[K[15H[K[16H[K[17H[K[18H[K[19H[K[20H[K[21H[K[22H[K[23H[K[24H[K[25H[K[26H[K[27H[K[28H[K[29H[K[30H[K[31H[K[32H[K[33H[K[34H[K[35H[K[36H[K[37H[K[1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:2  1 sel - client0@[13450] - Kakoune[1H[39mO[30;47mn[39;49m Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:2  [34m1 sel[36m - client0@[13450][1;2H]2;/tmp/aerc-compose-4230040883.eml 1:3  1 sel - client0@[13450] - Kakoune[1H[39mOn[30;47m [39;49mSat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:3  [34m1 sel[36m - client0@[13450][1;3H]2;/tmp/aerc-compose-4230040883.eml 1:4  1 sel - client0@[13450] - Kakoune[1H[39mOn [30;47mS[39;49mat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:4  [34m1 sel[36m - client0@[13450][1;4H]2;/tmp/aerc-compose-4230040883.eml 1:5  1 sel - client0@[13450] - Kakoune[1H[39mOn S[30;47ma[39;49mt Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:5  [34m1 sel[36m - client0@[13450][1;5H]2;/tmp/aerc-compose-4230040883.eml 1:6  1 sel - client0@[13450] - Kakoune[1H[39mOn Sa[30;47mt[39;49m Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:6  [34m1 sel[36m - client0@[13450][1;6H]2;/tmp/aerc-compose-4230040883.eml 1:7  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat[30;47m [39;49mOct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:7  [34m1 sel[36m - client0@[13450][1;7H]2;/tmp/aerc-compose-4230040883.eml 1:8  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat [30;47mO[39;49mct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:8  [34m1 sel[36m - client0@[13450][1;8H]2;/tmp/aerc-compose-4230040883.eml 1:9  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat O[30;47mc[39;49mt 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:9  [34m1 sel[36m - client0@[13450][1;9H]2;/tmp/aerc-compose-4230040883.eml 1:10  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oc[30;47mt[39;49m 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:10  [34m1 sel[36m - client0@[13450][1;10H]2;/tmp/aerc-compose-4230040883.eml 1:11  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oct[30;47m [39;49m15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:11  [34m1 sel[36m - client0@[13450][1;11H]2;/tmp/aerc-compose-4230040883.eml 1:12  1 sel - client0@[13450] - Kakoune
```

The second is one highlight char next (`1`) and is broken:

![screenshot broken](https://dav.missbanal.net/43ef3bd1-f991-4b19-abc5-c3dd8ac908df.png)

```
[?1049h[?1004h[>4;1m[>5u[22t[?25l=[?2026$p[?1006h[?1000h[?1002h[1H[K[2H[K[3H[K[4H[K[5H[K[6H[K[7H[K[8H[K[9H[K[10H[K[11H[K[12H[K[13H[K[14H[K[15H[K[16H[K[17H[K[18H[K[19H[K[20H[K[21H[K[22H[K[23H[K[24H[K[25H[K[26H[K[27H[K[28H[K[29H[K[30H[K[31H[K[32H[K[33H[K[34H[K[35H[K[36H[K[37H[K[1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:1  1 sel - client0@[13450] - Kakoune[m[1H[30;47mO[39;49mn Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[2H> Can be closed according to this comment https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-895548023[K[3H>[K[4H> -- [K[5H> Reply to this email directly or view it on GitHub:[K[6H> https://github.com/pixelfed/pixelfed/issues/2224#issuecomment-1279780732[K[7H> You are receiving this because you are subscribed to this thread.[K[8H>[K[9H> Message ID: <pixelfed/pixelfed/issues/2224/1279780732@github.com>[K[10H[K[11H[34m~[K[12H~[K[13H~[K[14H~[K[15H~[K[16H~[K[17H~[K[18H~[K[19H~[K[20H~[K[21H~[K[22H~[K[23H~[K[24H~[K[25H~[K[26H~[K[27H~[K[28H~[K[29H~[K[30H~[K[31H~[K[32H~[K[33H~[K[34H~[K[35H~[K[36H~[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:1  [34m1 sel[36m - client0@[13450][1;1H]2;/tmp/aerc-compose-4230040883.eml 1:2  1 sel - client0@[13450] - Kakoune[1H[39mO[30;47mn[39;49m Sat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:2  [34m1 sel[36m - client0@[13450][1;2H]2;/tmp/aerc-compose-4230040883.eml 1:3  1 sel - client0@[13450] - Kakoune[1H[39mOn[30;47m [39;49mSat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:3  [34m1 sel[36m - client0@[13450][1;3H]2;/tmp/aerc-compose-4230040883.eml 1:4  1 sel - client0@[13450] - Kakoune[1H[39mOn [30;47mS[39;49mat Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:4  [34m1 sel[36m - client0@[13450][1;4H]2;/tmp/aerc-compose-4230040883.eml 1:5  1 sel - client0@[13450] - Kakoune[1H[39mOn S[30;47ma[39;49mt Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:5  [34m1 sel[36m - client0@[13450][1;5H]2;/tmp/aerc-compose-4230040883.eml 1:6  1 sel - client0@[13450] - Kakoune[1H[39mOn Sa[30;47mt[39;49m Oct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:6  [34m1 sel[36m - client0@[13450][1;6H]2;/tmp/aerc-compose-4230040883.eml 1:7  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat[30;47m [39;49mOct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:7  [34m1 sel[36m - client0@[13450][1;7H]2;/tmp/aerc-compose-4230040883.eml 1:8  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat [30;47mO[39;49mct 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:8  [34m1 sel[36m - client0@[13450][1;8H]2;/tmp/aerc-compose-4230040883.eml 1:9  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat O[30;47mc[39;49mt 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[99C/tmp/aerc-compose-4230040883.eml 1:9  [34m1 sel[36m - client0@[13450][1;9H]2;/tmp/aerc-compose-4230040883.eml 1:10  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oc[30;47mt[39;49m 15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:10  [34m1 sel[36m - client0@[13450][1;10H]2;/tmp/aerc-compose-4230040883.eml 1:11  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oct[30;47m [39;49m15, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:11  [34m1 sel[36m - client0@[13450][1;11H]2;/tmp/aerc-compose-4230040883.eml 1:12  1 sel - client0@[13450] - Kakoune[1H[39mOn Sat Oct [30;47m1[39;49m5, 2022 at 6:44 PM CEST, MrAn0nym wrote:[K[37H[36m[K[98C/tmp/aerc-compose-4230040883.eml 1:12  [34m1 sel[36m - client0@[13450][1;12H]2;/tmp/aerc-compose-4230040883.eml 1:12  prompt - client0@[13450] - Kakoune
```

The diff is this last instruction. What is wrong with it ?

```
^G^[[1H^[[39mOn Sat Oct ^[[30;47m1^[[39;49m5, 2022 at 6:44 PM CEST, MrAn0nym wrote:^[[K^[[37H^[[36m^[[K^[[98C/tmp/aerc-compose-4230040883.eml 1:12  ^[[34m1 sel^[[36m - client0@[13450]^[[1;12H^[]2;/tmp/aerc-compose-4230040883.eml 1:12  prompt - client0@[13450] - Kakoune
```

**Comment by Screwtapello** [comment]
That's not just one instruction, but several. Breaking it down:

- `^G` is the tail of the previous instruction, setting the terminal title text
- `^[[1H` moves the cursor to the top left of the screen
- `^[[39m` set the text foreground to the terminal's default colour
- `On Sat Oct ` prints text
- `^[[30;47m` sets the text to black on white
- `1` prints text
- `^[[39;49m` sets the text foreground and background to defaults
- `5, 2022 at 6:44 PM CEST, MrAn0nym wrote:` prints text
- `^[[K` clears all the text to the end of the line
- `^[[37H` moves the cursor to line 37, column 1
- `^[[36m` sets the foreground colour to cyan
- `^[[K` clears all the text to the end of the line
- `^[[98C` moves the cursor right 97 cells
- `/tmp/aerc-compose-4230040883.eml 1:12  ` prints text
- `^[[34m` sets the foreground colour to blue
- `1 sel` prints text
- `^[[36m` sets the foreground colour to cyan again
- ` - client0@[13450]` prints text
- `^[[1;12H` moves the cursor to line 1, column 12
- `^[]2;/tmp/aerc-compose-4230040883.eml 1:12  prompt - client0@[13450] - Kakoune` is the head of the next instruction, setting the terminal title text again

None of that looks very suspicious to me, so I don't think the probelem is there.

Looking at your screenshots, though, it looks like there's a bunch of extra text that doesn't appear to be related to what Kakoune (or aerc) are drawing. There's some blanks in the line of `~`s down the left-hand edge, the text `ga0@l` near the bottom, and something about `akouneget_string:`. There's no `get_string` function in the Kakoune codebase; is it possible there's some other app or tool running in the background and writing warnings to the terminal Kakoune is running in?

**Comment by Screwtapello** [comment]
Also, at this point I'm fairly sure the problem @0x5424 is seeing is unrelated. These same diagnosis steps might help figure out what the problem is, but if it's a problem with Kakoune, it should probably be filed as a separate issue.

**Comment by stacyharper** [comment]
> None of that looks very suspicious to me, so I don't think the probelem is there.

I also didnt figured out what is wrong with this instruction. But it is the only difference between the two screenshot results. I tested those two scripts using a fake editor as aerc editor. This fake editor only output the typescripts and sleep for some time for me to inspect the result in aerc. There is no Kakouneget_string in the typescripts and I dont think this is something from aerc itself. So I'm very confused also…

**Comment by stacyharper** [comment]
Here a reproduction script. Configure it as aerc editor and try to reply to a mail.

https://paste.sr.ht/~stacyharper/f7607bcad81b9380666162bfcd3ba53e57705a45

**Comment by stacyharper** [comment]
Is it possible something is wrong in those typescripts ? May we be writing outside the box ? Or maybe go-libvterm got a problem. I think the get_string comes from here

**Comment by Screwtapello** [comment]
It's quite possible go-libvterm has a problem, terminal emulation has a lot of details and corner-cases. I don't know much about aerc or Go, but it [looks like](https://git.sr.ht/~rjarry/aerc/tree/master/item/go.mod) the latest version of aerc uses a library called [tcell-term](https://sr.ht/~rockorager/tcell-term/) instead? Hopefully that behaves a little more reliably.

**Comment by stacyharper** [comment]
Ya I was discussing with aerc folks on irc and they told me they switched this library. I builded aerc master and the issue seems gone. I'll see after some time if some corner cases remains. I'll comment or close this issue accordingly !
